Fetch upstream from halide/Halide to InteonCo/Halide:master

DawnStone opened the pull request.

dawn-inteon commented:
Closing - this change was never agreed for merge - but we can keep the branch if desired.

steven-johnson committed:
Silence "may be used uninitialized" in Buffer::for_each_element() (#6747)

In at least one version of GCC (Debian 11.2.0-16+build1), an optimized build using `Buffer::for_each_element(int *pos)` will give (incorrect) compiler warnings/errors that "pos may be used uninitialized).

From inspection of the code I feel pretty sure this is a false positive -- i.e., the optimizer is confused -- and since no other compiler we've encountered issues a similar warning (nor do we see actual misbehavior), I'm inclined not to worry -- but the warning does break some build configurations.

Rather than try to fight with selectively disabling this warning, I'm going to propose inserting a memset() here to reassure the compiler that the memory really is initialized; while it's unnnecessary, it's likely to be insignificant compared to the cost of usual calls to for_each_element().

(BTW, this is not a new issue, I've seen it for quite a while as this GCC is the default on one of my Linux machines... it just finally annoyed me enough to want to make it shut up.)

steven-johnson committed:
Update WABT to 1.0.29 (#6748)

steven-johnson committed:
Update hannk README link to hosted models page (#6749)

The current one is being sunsetted

steven-johnson committed:
Add a `HalideError` base class to Python bindings (#6750)

* Add a `HalideError` base class to Python bindings

Per suggestion from @alexreinking, this remaps all exceptions thrown by the Halide Python bindings to be `halide.HalideError` (or a subclass thereof), rather than plain old `RuntimeError`.

* Remove scalpel left in patient

* Don't use a subclass for PyStub error handling

steven-johnson committed:
Deprecate GeneratorContext getters with `get_` prefix (#6753)

Minor hygiene: most getters in Halide don't have a `get_` prefix. These are very rarely used (only one instance in our test suite I could find) but, hey, cleanliness.

steven-johnson committed:
Add GeneratorFactoryProvider to generate_filter_main() (#6755)

* Add GeneratorFactoryProvider to generate_filter_main()

This provides hooks to allow overriding the Generator(s) that generate_filter_main() can use; normally it defaults to the global registry of C++ Generators, but this allows for (e.g.) alternate-language-bindings to selectively override this (e.g. to enumerate only Generators that are visible in that language, etc).

(No visible change in behavior from this PR; this is just cherry-picked from work-in-progress elsewhere to simplify review & merge)

* Update Generator.cpp

* Fix error handling

steven-johnson committed:
Deprecate disable_llvm_loop_opt (#4113) (#6754)

This PR proposes to (finally) deprecate disable_llvm_loop_opt:
- make LLVM codegen default to no loop optimization; you must use enable_llvm_loop_opt explicitly to enable it
- disable_llvm_loop_opt still exists, but does nothing (except issue a user_warning that the feature is deprecated)
- Remove various uses of disable_llvm_loop_opt
- Add comments everywhere that the default is different in Halide 15 and that the disable_llvm_loop_opt feature will be removed entirely in Halide 16

Note that all Halide code at Google has defaulted to having disable_llvm_loop_opt set for ~years now, so this is a well-tested codepath, and consensus on the Issue seemed to be that this was a good move.

steven-johnson committed:
Minor metadata-related cleanups (#6759)

(Harvested from #6757, which probably won't land)

- Add clarifying comment/reference in Generator
- Add assertion to compile_to_multitarget() function
- Fix misleading/wrong code in correctness_compile_to_multitarget

steven-johnson committed:
Expand the x86 SIMD variants tested in correctness_vector_reductions (#6762)

A recent bug in LLVM codegen was missed because it only affected x86 architectures with earlier-than-AVX2 SIMD enabled; it didn't show up for AVX2 or later. This revamps correctness_vector_reductions to re-run multiple times when multiple SIMD architectures are available on x86 systems.

(correctness_vector_reductions was chosen here because it reliably demonstrated the specific failures in this case.)

steven-johnson committed:
Fix Param<T>::set_estimate for T=void (#6766)

* Fix Param<T>::set_estimate for T=void

* Add tests

steven-johnson committed:
add_python_aot_extension should use FUNCTION_NAME for the .so output … (#6767)

add_python_aot_extension should use FUNCTION_NAME for the .so output (otherwise you can't produce multiple aot extensions from the same Generator)

vksnk committed:
Update the list of fused_pairs and run validate_fused_group for specalization definitions too (#6770)

* Update the list of fused_pairs and run validate_fused_group for specialization definitions too.

Fixes https://github.com/halide/Halide/issues/6763.

* Address review comments

* Add const to auto&

steven-johnson committed:
Add Func::type()/types(), deprecate Func::output_type()/output_types() (#6772)

* rename GIOBase::type() and friends

* Func::output_type() -> Func::type()

* Add type() forwarders for inputs

* Add Func::dimensions() wrapper

* Update Func.h

LebedevRI committed:
Fix fundamental confusion about target/tune CPU (#6765)

* Fix fundamental confusion about target/tune CPU

Sooo. Uh, remember when in https://github.com/halide/Halide/pull/6655
we've agreed that we want to add support to precisely specify
the CPU for which the code should be *tuned* for,
but not *targeted* for. Aka, similar to clang's `-mtune=` option,
that does not affect the ISA set selection?

So guess what, that's not what we did, apparently.
`CodeGen_LLVM::mcpu()` / `halide_mcpu` actually do specify
the *target* CPU. It was obvious in retrospect, because e.g.
`CodeGen_X86::mattrs()` does not, in fact, ever specify `+avx2`,
yet we get AVX2 :) So we've unintentionally added `-march=` support.
Oops.

While i'd like to add `-march=` support, that was not the goal here.

Fixing this is complicated by the fact that
`llvm::Target::createTargetMachine()` only takes `CPU Target` string,
you can't specify `CPU Tune`.

But this is actually a blessing in disguise,
because it allows us to fix another bug at the same time:

There is a problem with halide "compile to llvm ir assembly",
a lot of information from Halide Target is not //really// lowered
into LLVM Module, but is embedded as a metadata,
that is then extracted by halide `make_target_machine()`.

While that is not a problem in itself, it makes it *impossible*
to dump the LLVM IR, and manually play with it,
because e.g. the CPU [Target] and Attributes (ISA set)
are not actually lowered into the form LLVM understands,
but are in some halide-specific metadata.

So, to fix the first bug, we must lower the CPU Tune
into per-function `"tune-cpu"` metadata,
and while there we might as well lower `"target-cpu"`
and `"target-features"` similarly.

* Address review notes

* Hopefully silence bogus issue reported by ancient GCC

* Call `set_function_attributes_from_halide_target_options()` when JIT compiling

* Fix grammar

steven-johnson committed:
Fix annoying typo in Func.h (#6774)

Update Func.h

steven-johnson committed:
Add execute_generator() API (#6771)

This refactors the existing `generate_filter_main()` call in two, moving the interesting implementation of how to drive AOT into the new `execute_generator()` call (reducing `generate_filter_main()` to parsing argc/argv and error reporting).

The new `execute_generator()` is intended to be used (eventually) from Python, as a way to drive Generator compilation from a Python script more easily. The PR doesn't provide a Python wrapper for this call yet (that will come in a subsequent PR).

Also, a drive-by removal of the "error_output" arg to generate_filter_main() -- AFAICT, no one has ever used it for anything but stderr, and the refactoring now just directs all errors to `user_error` uniformly.

steven-johnson committed:
Allow overriding of `Generator::init_from_context()` for debug purposes (#6760)

* Allow overriding of `Generator::init_from_context()` for debug purposes

* Update Generator.h

* Attempt to clarify contract

steven-johnson committed:
Convert some assert-only usage of output_types() -> types() (#6779)

LebedevRI committed:
[miscompile] Don't de-negate and change direction of shifts-by-unsigned (#6782)

I'm afraid the problem is really obvious:
https://github.com/halide/Halide/blob/b5f024fa83b6f1cfe5e83a459c9378b7c5bf096d/src/CodeGen_LLVM.cpp#L2628-L2649
^ the shift direction is treated as flippable by the codegen
iff the shift amount is signed :)

The newly-added test fails without the fix.

I've hit this when writing tests for https://github.com/halide/Halide/pull/6775

steven-johnson committed:
Move some options from execute_generator back to generate_filter_main (#6787)

Loading plugins and setting the default autoscheduler name both change global state, which isn't a desirable fit for execute_generator(), since it's not intended to mutate global state. (Mutating the state from a main() function is of course a reasonable thing to do.)

LebedevRI committed:
LLVM codegen: register AA pipeline if LLVM is older than 14 (#6785)

It's the default after https://reviews.llvm.org/D113210 / https://github.com/llvm/llvm-project/commit/13317286f8298eb3bafa9ddebd1c03bef4918948,
but still needs to be done for earlier LLVM's.

Refs. https://github.com/halide/Halide/issues/6783
Refs. https://github.com/halide/Halide/pull/6718

Partially reverts https://github.com/halide/Halide/pull/6718

steven-johnson committed:
halide_type_of<>() should always be constexpr (#6790)

The ones in HalideRuntime.h have been marked constexpr for a while, but the ones in Float16.h got missed

steven-johnson committed:
Define an AbstractGenerator interface (#6637)

* AbstractGenerator (rebased, v3)

* Update AbstractGenerator.h

* clang-format

* Update Generator.cpp

* IOKind -> ArgInfoKind

* Various cleanups of AbstractGenerator

* clang-format

* fix pystub

* Update abstractgeneratortest_generator.cpp

* dead code

* ArgInfoDirection

* cleanup

* Delete PyGenerator.cpp

* Update PyStubImpl.cpp

* Update PyStubImpl.cpp

* Fixes from review comments

* Remove `get_` prefix from getters in AbstractGenerator

* Missed some fixes

* Fixes

* Add GeneratorFactoryProvider for generate_filter_main()

* Add GeneratorFactoryProvider to generate_filter_main()

This provides hooks to allow overriding the Generator(s) that generate_filter_main() can use; normally it defaults to the global registry of C++ Generators, but this allows for (e.g.) alternate-language-bindings to selectively override this (e.g. to enumerate only Generators that are visible in that language, etc).

(No visible change in behavior from this PR; this is just cherry-picked from work-in-progress elsewhere to simplify review & merge)

* Update Generator.cpp

* fixes

* Update Generator.cpp

* Restore build_module() and build_gradient_module() methods

* Update Generator.h

* fixes

* Update Generator.cpp

* Update AbstractGenerator.h

steven-johnson committed:
hexagon_scatter test should run only if target has HVX (#6793)

It will run otherwise, but is slow on some other targets; rather than trying to (e.g.) shard it, just skip it

Commit:
Add Target support for architectures with implementation specific vector size. (#6786)

Move vector_bits_* Target support from fixed_width_vectors branch to
make smaller PRs.

steven-johnson committed:
slow tests should support sharding (#6780)

* slow tests should support sharding

The simd_op_check test suite is pretty slow (especially for wasm, where it is interpreted); at one point we tried to use ThreadPool to speed it up, but too many pieces of Halide IR aren't threadsafe and we disabled it long ago.

This removes the ThreadPool usage entirely, and instead adds support for the GoogleTest 'sharded test' protocol, which uses certain env vars to allow a test to opt in for splitting its test into smaller pieces.

At present our buildbot isn't attempting to make use of this feature, but it will be a big win for downstream usage in Google, where tests that run "too long" are problematic and splitting them into multiple shards makes various day to day activiites much more pleasant.

steven-johnson committed:
Add missing include to test_sharding.h (#6795)

steven-johnson committed:
Pacify clang-tidy (#6796)

* Pacify clang-tidy

Newer versions can warn about "parameter 'f' shadows member inherited from type 'StubOutputBufferBase'", etc -- easy enough

* Update .clang-tidy

steven-johnson committed:
Silence a "possibly uninitialized" warning (#6797)

* Silence a "possibly uninitialized" warning

At least one compiler thinks we can use this without initialization, which isn't true, but this silences it.

* trigger buildbots

steven-johnson committed:
Make all tests default to `-fvisibility=hidden` (#6799)

* Step 1

* still more

* Export the error classes so they can be caught

steven-johnson committed:
Minor typedef cleanup (#6800)

* cleanup

* format

steven-johnson committed:
Fix auto_schedule/machine_params parsing (#6804)

The recent refactoring that added `execute_generator` accidentally nuked setting these two GeneratorParams. Oops. Fixed.

vksnk committed:
Rewrite strided loads of 4 in AlignLoads (#6806)

* Rewrite strided loads of 4 in AlignLoads

* Add a check for strided 4 load

abadams committed:
Fix two minor bugs triggered by an or reduction with early-out (#6807)

* Fix two minor bugs triggered by a or reduction with early-out

* Gotta print success

* Appease clang-tidy

LebedevRI committed:
[CMake] Mark multi-threaded tests as such (#6810)

Commit:
Add support for vscale vector code generation. (#6802)

Add support for vscale vector code generation. Factored from the
fixed_length_vectors branch to make PRs smaller and easier to review.

This will be used to support the ARM SVE/SVE2 and RISC V Vector
architectures.

alexreinking committed:
Rework .gitignore (#6822)

* reorganize .gitignore

* Add exclusions for CMake build

* .gitignore: comment, drop stale rules

* fully and precisely exclude CMake build tree

* add debugging directions to .gitignore

* ignore CMake install tree

* Sort groups

alexreinking committed:
Update presets to format version 3 (#6824)

steven-johnson committed:
Fix for top-of-tree LLVM (#6825)

steven-johnson committed:
Tweak python apps for better Blaze/Bazel compatibility (#6823)

* Tweak python apps/tutorials for better Blaze/Bazel compatibility

- Don't write to current directory (rely on an env var to say where to write)
- Don't read from arbitrary absolute paths (again, rely on an env var)
- Drive-by removal of unnecessary #include in Codegen_LLVM.cpp inside a lambda (!)

* Recommended fixes

* Revert all changes to tutorial

* Revise apps

* Remove apps_helpers.py

steven-johnson committed:
Change stub module names in Python to be _pystub rather than _stub (#6830)

This is a bit finicky, but making this the default nomenclature will make some downstream usages less ambiguous and a bit easier to manage. (Yes, I realize that #6821 removes the Makefile entirely, but until it lands, it needs fixing there too.)

alexreinking committed:
Apply CMAKE_C_COMPILER_LAUNCHER to initmod clang calls (#6831)

alexreinking committed:
Remove Python bindings from Makefiles (#6821)

* Remove Python bindings from Makefiles

* Restore test_li2018 in Makefile (now C++-only)

* Add dummy `test_python` target for buildbots

steven-johnson committed:
Add a new, alternate JIT-call convention (#6777)

* Prototype of revised JIT-call convention

Experiment to try out a way to call JIT code in C++ using the same calling conventions as AOT code. Very much experimental.

* Update Pipeline.h

* Add Python support for `compile_to_callable` + make empty_ucon static

* Update PyCallable.cpp

* Update buffer.py

* wip

* Update callable.py

* WIP

* Update custom_allocator.cpp

* Update Callable.cpp

* Add Generator support for Callables

* Update Generator.cpp

* Update PyPipeline.cpp

* Fixes

* Update callable.cpp

* Update CMakeLists.txt

* create_callable_from_generator

* More cleanup

* Update Generator.cpp

* Fix Python bounds inference

* Add Python wrapper for create_callable_from_generator() + Add kwarg support for Callable

* Add set_generatorparam_values() + usage

* Fix auto_schedule/machine_params parsing

The recent refactoring that added `execute_generator` accidentally nuked setting these two GeneratorParams. Oops. Fixed.

* Move the type-checking code into a constexpr code

* Update Callable.h

* clang-tidy

* CLANG-TIDY

* Add `make_std_function`, + more general cleanup

* Update example_jittest.cpp

* Update Callable.h

* Update Callable.h

* More tweaking, smaller CallCheckInfo

* Still more cleanup

* make_std_function now does Buffer type/dim checking where possible

* Add tests for calling `AbstractGenreator::compile_to_callable()` directly

* enable exports

* Various fixes

* Improve fill_slot for Halide::Buffer

* kill report_if_error

* Update callable_bad_arguments.cpp

* Update Pipeline.cpp

* Revise error handling

* Update Callable.cpp

* Update callable.py

* Update callable_generator.cpp

* Update callable.py

* HALIDE_MUST_USE_RESULT -> HALIDE_FUNCTION_ATTRS for Callable

alexreinking committed:
Scrub Python from Makefile after buildbot update (#6833)

steven-johnson committed:
Remove unused function in callable_generator.cpp (#6834)

steven-johnson committed:
Disable testing for apps/linear_algebra on x86-32-linux/Make (#6836)

* Disable testing for apps/linear_algebra on x86-32-linux/Make

This wasn't biting us before because we were disabling *all* apps/ on x86-32-linux (oops); the recent change to remove python testing under Make also re-enabled this test.

TL;DR: this can probably be made to work somehow, but it's not worth debugging, since that case is both pretty nice, and already covered under CMake. It's literally not worth the time to fix.

* Update Makefile

steven-johnson committed:
Rearrange subdirectories in python_bindings (#6835)

This is intended to facilitate a few things:
- Move all Generators used in tests, apps, etc to a single directory to simplify the build rules (this is especially useful for the work in https://github.com/halide/Halide/pull/6764)
- Put all the test and apps stuff under a single directory to facilitate adding some Python packaging that can make integration into Bazel/Blaze builds a bit less painful

@alexreinking, does this look like the layout we discussed before?

abadams committed:
Better lowering of halving_sub and rounding_halving_add (#6827)

* Better lowering of halving_sub and rounding_halving_add

Previously, lower_halving_sub and lower_rounding_halving_add both used 9
ops. This change redirects halving_sub to use rounding_halving_add, and
redirects rounding_halving_add to use halving_add. In the case that none
of these instructions exist natively, this reduces it to 7/8 ops for
signed/unsigned halving sub and 6 ops for rounding halving add.

More importantly, this lets halving_sub make use of pavgw/b on x86 to
reduce it to 3 ops for u8 and u16 inputs.

* Make signed rounding_halving_add on x86 use pavgb/w too

* Cast result back to signed

* Add explanatory comment

* Fix comment

* Add explanation of signed case

alexreinking committed:
Check RDom::where predicates for race conditions (#6842)

Fixes #6808

steven-johnson committed:
Remove Generator::value_tracker and friends (#6845)

This is an internal-to-Generator helper that is used to try to detect certain classes of errors when using GeneratorStubs. To the best of my knowledge, it has ~never found a useful error in all of its existence; combined with the very limited usage of GeneratorStubs, I think this code no longer pays for itself, and should be removed.

(Note that this was never externally visible, thus no deprecation warnings should be necessary.)

steven-johnson committed:
Deprecate/remove Generator::get_externs_map() and friends (#6844)

* Deprecate/remove Generator::get_externs_map() and friends

This is a feature of Generator that was added years ago to allow adding external code libraries in LLVM bitcode form (rather than simply as extern "C" or similar). In theory it allow for better codegen for external code modules (since LLVM has access to all the bitcode for optimization); in practice, we only know of one project that ever used it, and that project no longer exists. Additionally, it tended to be fairly flaky in terms of actual use -- e.g., missing symbols tended to crop up unpredictably.

The issues with this feature are likely fixable, but since it hasn't (AFAICT) been used in ~years, we're better off deprecating it for Halide 15 and removing for Halide 16.

(If anyone out there is still relying on this feature, obviously you should speak up ASAP.)

* Also remove ExternalCode.h & friends

* Also remove correctness/external_code.cpp

* HALIDE_ALLOW_GENERATOR_EXTERNS_MAP -> HALIDE_ALLOW_GENERATOR_EXTERNAL_CODE

steven-johnson committed:
Add placeholder code for bfloat16 in Python (#6849) (#6850)

* Add placeholder code for bfloat16 in Python (#6849)

This is a no-op change; I just want to mark the place(s) in the Python bindings that need attention if/when it becomes possible to support bfloat16 in Python buffers.

* Update PyBinaryOperators.h

steven-johnson committed:
Fix the PLUGINS argument to properly join multiple arguments (#6851)

steven-johnson committed:
Add autoscheduling to the generator_aot_stubuser test (#6855)

* Add autoscheduling to the generator_aot_stubuser test

* fix test_apps

* fix test_apps, again

steven-johnson committed:
Silence Adams2019 Autoscheduler (#6854)

* Make aslog() a proper ostream

* Ensure that all `dump()` calls take and use an ostream

* Progress Bar only draws as LogLevel >= 1

* clang-format

* Rework all aslog(0) statements

* Update ASLog.cpp

* syntax

* Update ASLog.cpp

* Revert fancy aslog stuff

* Update ASLog.h

* trigger buildbots

steven-johnson committed:
Rework autoscheduler API (#6788) (#6838)

* Rework autoschduler API (#6788)

* Oops

* Update test_function_dag.cpp

* clang-tidy

* trigger buildbots

* Update Generator.h

* Minor cleanups

* Update README_cmake.md

* Check for malformed autoscheduler_params dicts

* Add alias-with-autoscheduler code, plus tweaks

* Update stubtest_jittest.cpp

* Update Makefile

* trigger buildbots

* fixes

* Update AbstractGenerator.cpp

* Update stubtest_generator.cpp

* Update Makefile

* Add deprecation warning for HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API

* Make AutoschedulerParams a real struct

* clang-tidy

derek-gerstmann committed:
[vulkan phase0] Add adts for containers and memory allocation to runtime (#6829)

* Cherry pick runtime internals as standalone commit (preparation work for Vulkan runtime)

* Clang format/tidy fixes

* Fix runtime test linkage and include paths to not include libHalide

* Update test/runtime/CMakeLists.txt

Fix typo mismatch for HALIDE_VERSION_PATCH

Co-authored-by: Alex Reinking <reinking@google.com>

* Add compiler id guard to build options for runtime tests

* Avoid building runtime tests on MSVC since Halide runtime headers are not MS compatible
Remove CLANG warning flag for runtime test

* Change runtime test compile definitions to be PRIVATE.
Remove PUBLIC_EXPORTS from runtime test definition.

* Add comment about GNU warnings for 'no-builtin-declaration-mismatch'

* Change to debug(user_context) for debug messages where context is valid.
Wrap verbose debugging with DEBUG_RUNTIME ifdef.
Syle pass based on review comments.

* Add note explaining why we disable the internal runtime tests on MSVC.

* Cleanup cmake logic for disabling runtime internal tests for MSVC and add a status message.

* Don't use strncpy for prepend since some implementations may insert a null char regardless of the length used

* Workaround varying platform str implementations and handle termination
directly.

* Clang Tidy/Format pass

Co-authored-by: Derek Gerstmann <dgerstmann@adobe.com>
Co-authored-by: Alex Reinking <reinking@google.com>

LebedevRI committed:
Promote Reinterpret Intrinsic into an Reinterpret IR Node (#6853)

* Promote Reinterpret Intrinsic into an Reinterpret IR Node

As discussed in https://github.com/halide/Halide/issues/6801#issuecomment-1152731683

I don't think this is complete, there are likely a few more places
that need to be taught about it still, altough i think this is mostly it.

Note that this only promotes the intrinsic,
this does not adjust it's handling, as hinted in:
https://github.com/halide/Halide/issues/6801#issuecomment-1155603752

* Silence buildbot warning

* Speculative fix for Codegen C failure?

* Restore comment

* Delete obsolete FIXME

* RegionCost: reinterpret is free

* LICM: actually adjust the comment

alexreinking committed:
Python source reorg (#6867)

* Move python binding sources to src/halide/halide_

* Rename native module to halide_

* Fix tests

* Avoid copying Python sources

* Fix installation rules

* Make diff smaller

* trigger buildbots

* Add issue todo

Co-authored-by: Steven Johnson <srj@google.com>

steven-johnson committed:
Fix simd_op_check for top-of-tree LLVM (#6874)

* Fix simd_op_check for top-of-tree LLVM

* clang-format

rootjalex committed:
Use pmaddubsw 8-bit horizontal widening adds (Fixes #6859) (#6873)

* use pmaddubsw 8-bit horizontal widening adds

* add SSE3 versions too

* add pmaddubsw tests

LebedevRI committed:
[Codegen_LLVM] Radically simplify `visit(const Reinterpret *op)` (#6865)

1. LLVM IR `bitcast` happily bitcasts between vectors and scalars: https://godbolt.org/z/9zqx11rna
2. `ptrtoint` already implicitly truncates/zero-extends if the int is larger than the pointer type: https://llvm.org/docs/LangRef.html#ptrtoint-to-instruction
3. `inttoptr` already implicitly truncates/zero-extends if the int is larger than the pointer type: https://llvm.org/docs/LangRef.html#inttoptr-to-instruction

So we don't need to do any of that 'special' handling.

LebedevRI committed:
[Codegen] Fail to codegen `Call::undef`, just like `Call::signed_integer_overflow` (#6871)

See discussion in https://github.com/halide/Halide/pull/6866.
It's not obvious if that codepath is ever hit,
let's optimistically assume that it is not.

If this turns out to be not true, we'll have to deal with
a more complicated question of the proper lowering for it,
can it be `poison`, or must it be a `freeze poison`.

steven-johnson committed:
Fix error in Makefile for Adams2019 on OSX (#6877)

We erroneously link in the dylib and also dynamically load it, causing an error. We should skip the linkage and always load dynamically..

steven-johnson committed:
Refactor/cleanup in Autoscheduler code (#6858)

* Move ASLog.cpp/.h to common/

* Add trivial Parsing utility & use it

* Update ParamParser.h

* fixes

* fixes

shoaibkamil committed:
Ensure $CMAKE_{lang}_OUTPUT_EXTENSION is set before using it (#6879)

Ensure CMAKE_{lang}_OUTPUT_EXTENSION is set before using it

Co-authored-by: Shoaib Kamil <kamil@adobe.com>

derek-gerstmann committed:
#6863 - Fixes to make address sanitizer happy for internal runtime classes (#6880)

* Fixes to make address sanitizer happy.
Fixed initialization defects in StringStorage that could cause buffer overruns
Fixed memory leaks within RegionAllocator and BlockAllocator
Added system memory allocation tracking to all internal runtime tests.

* Clang Tidy / Format pass

* Fix formatting to use braces around if statements

Co-authored-by: Derek Gerstmann <dgerstmann@adobe.com>

LebedevRI committed:
[Codegen_LLVM] Define all the things (#6866)

Long-term plan for LLVM is to get rid of `undef`,
and replace it with zero-initialization,
err, `poison`, because it has nicer semantics.

Everywhere we use `undef` as a placeholder in shuffle
(be it either for a second operand, or undef shuffle mask element),
or as a base 'empty' vector we are about to fully override
via insertelement, we can just switch those to poison nowadays.

The scary part is the `Call::undef` semantics/lowering,
perhaps it will need to be `freeze poison`.

steven-johnson committed:
Add set-host-dirty/copy-to-host to PythonExtensionGen (#6869)

* Add set-host-dirty/copy-to-host to PythonExtensionGen

See https://github.com/halide/Halide/issues/6868: Python Buffers are host-memory-only, so if the AOT-compiled halide code runs on (say) GPU, it may fail to copy the inputs to device and/or the results back to host. This fixes that. (We still need a solution that allows for lazy copies, but that will require adding another protocol that supports it.)

* Update PythonExtensionGen.cpp

steven-johnson committed:
Rewrite PythonExtensionGen to be C++ based (#6888)

* Rewrite PythonExtensionGen to be C++ based

This is intended as an alternative to #6885 -- this is even *more* gratuitous, but:
- We have ~always compiled Python extensions using C++ anyway
- This code is arguably terser, cleaner, and safer (the cleanups happen via dtors)
- The code size difference is negligible (~300 bytes out of 160k for addconstant.cpython-39-darwin.so)

* Update PythonExtensionGen.cpp

steven-johnson committed:
Fixes to allow compiling with LLVM16 (#6889)

abadams committed:
Add support for generating x86 sum-of-absolute-difference reductions (#6872)

steven-johnson committed:
Remove (most) of the env var usage from Adams2019 (#6861)

* Move ASLog.cpp/.h to common/

* Add trivial Parsing utility & use it

* Update ParamParser.h

* fixes

* wip

* fixes

* Fixes

* clang-format

* Update Makefile

* Remove may_subtile

* Update Cache.cpp

* Update Cache.cpp

* Update AutoSchedule.cpp

* Update AutoSchedule.cpp

derek-gerstmann committed:
[vulkan phase1] Add SPIR-V IR  (#6882)

* Import SPIRV-IR from personal branch

* Refactor SPIR-V IR into separate header / source files.

* Refactory SPIR-V factory methods.
Fix SPIR-V interface library and header paths.
Add SPIR-V internal test.

* Hookup internal SPIRV IR test

* Fixes and cleanups to address PR #6882
Refactor logic of SPIR-V dependency to make fetch dependecy optional
Change SPIR-V fetch dependency to avoid building and just populate contents
Change SPIR-V internal test to always link against method ... only enabled if WITH_SPIRV is defined
Add missing SPIRV target feature

* Update src/CMakeLists.txt

Co-authored-by: Alex Reinking <reinking@google.com>

* Add missing iostream header when WITH_SPIRV is undefined

* Fix declaration ordering for TARGET_SPIRV option so that dependencies get triggered

* Turn on FETCH_SPIRV_HEADERS by default to get build to pass for now

* Fix path finding logic for SPIR-V header path from populated fetch dependency

* Revert back to Halide_SPIRV target name

* Don't use imported interface for SPIR-V.
Use Halide_SPIRV naming since target is defined before Halide itself.

* Add local copy of SPIR-V header file, along with license and readme.
Update CMake rules to use local include path by default.

* Make SPIR-V include path a system path to avoid clang format/tidy processing

* Remove SpirvIR.h header file from being included with Halide.h (since it's only used internally for CodeGen)

* Add ./dependencies/spirv to clang format ignore file

* Add comment about *not* including internally used headers like SpirvIR.h

* Refactor is_defined() asserts into check_defined() for reuse

* Add comment to SpirvIR.h header clarifying this file should not be exported.
Fix formatting to avoid single line if statements.
Use reserve for constructing vector components

* Rename hash_* methods to make_*_key methods (since they construct a key and don't actually hash the value)
Fix typo on components

* Clang format/tidy pass

* Fix formatting for more single-line if statements

* Disable TARGET_SPIRV by default for now

Co-authored-by: Derek Gerstmann <dgerstmann@adobe.com>
Co-authored-by: Alex Reinking <reinking@google.com>

steven-johnson committed:
Add `auto_schedule` label to Adams2019 and Li2018 tests in CMake (#6898)

* Add `auto_schedule` label to Adams2019 and Li2018 tests in CMake

These were ~never getting tested on the buildbots (and still aren't, I need to update it to run `auto_schedule` tests) but conceptually these tests should be in the same group as for Mullapudi.

Also, drive-by fix to broken test_apps_autoscheduler injected in https://github.com/halide/Halide/pull/6861.

* trigger buildbots

LebedevRI committed:
[Simplify] Drop no-op single-input identity shuffles (#6901)

LebedevRI committed:
[Codegen_LLVM] Annotate LLVM IR functions with `nounwind`/`mustprogress` attributes (#6897)

My reasoning is as follows, please correct me if i'm wrong:
1. Halide-generated code never throws exceptions
2. Halide-generated code always `call`s (as opposed to `invoke`s) the functions, there is no exception-safety RAII
3. Halide loops are meant to have finite number of iterations, they aren't meant to be endless and side-effect free
4. Halide (IR) assertions *might* abort.
5. Likewise, external callees *might* abort. (???)

Therefore, when not in presence of external calls,
it is obvious that (1) no exception will be unwinded
out of the halide-generated function, (2) none of the loops
will end up being endless with no observable side-effects.
... which is the semantics that is being stated by the
    LLVM IR function attributes `nounwind`+`mustprogress`.

I'm less clear as to what are the prerequisites on the behavior of
the external callees, but i do believe that they must also
at least not unwind. I guess they are also at least required to
either return or abort eventually.

rootjalex committed:
Don't try to fold saturating_sub of VectorReduce (#6896)

don't fold saturating_sub of VectorReduce

steven-johnson committed:
Upgrade clang-format and clang-tidy to v14 (v2) (#6902)

frengels committed:
Allow AMX instructions with K dimension larger than 4 bytes (#6582)

* recognize the patterns used for the RHS matrix

* make 1d tile matcher more robust

* put getting rhs tile's index into a separate func

* expand the tests used in correctness check

* add exclamation mark

* remove unused vars

* run format and tidy

* check for null before using IR in the next step

* check if the broadcast was found

* llvm below 13 is no longer supported

* replace single pattern with commutative permutations

* check if the stride is an `IntImm`, otherwise reject pattern

* apply clang-format-13

* rename wild_i32 -> v2

* check if v1 could be the stride value

* add more detail to a receiving a bad type

* added short explanation of the right-hand matrix layout

* added explanation for where the 4 comes from

* provide further documentation as to the layout of AMX

* add comments for expected patterns to get_3d_rhs_tile_index

* Document the matched pattern

Co-authored-by: Steven Johnson <srj@google.com>

abadams committed:
Fix autoscheduling trivial lut wrappers (#6905)

* Fix autoscheduling trivial lut wrappers

Fixes #6899

* trigger buildbots

Co-authored-by: Steven Johnson <srj@google.com>

steven-johnson committed:
Fix broken Makefile rules for autoschedulers on OSX (#6906)

* Fix broken Makefile rules for autoschedulers on OSX

A few issues here:
- Make was building the plugins as .dylib on OSX, but they should have been .so to match Linux (and just on general principles)
- On OSX, explicitly linking libHalide.dylib into a plugin means that it will load its own copy of libHalide, which is bad, because it means the plugin doesn't share the same set of globals. We need to omit that explicit dependency and allow it to just find the exported symbols at load time.
- Add a test to verify the fix; run it everywhere even though it should only have been failing for Make-build OSX builds.

Finally, let me add that we really need to set a sunset date for supporting Make in Halide. The Makefiles aren't really maintained properly anymore, and when something subtle goes wrong, it takes an unreasonable amount of time to debug for something that is no longer our canonical build tool.

* Use order-only prerequisites

* Remove new load_plugin.cpp test

Not worth the complexity for the extra test coverage.

alexreinking committed:
Start developing pip package (#6886)

Co-authored-by: Lukas Trümper <lukas.truemper@outlook.de>

steven-johnson committed:
LICENSE.txt: Include full text of Apache 2.0 license (not just the 'header' version) (#6912)

steven-johnson committed:
LICENSE.txt: add spirv license (#6913)

steven-johnson committed:
LICENSE.txt: add BLAS license. (#6914)

steven-johnson committed:
Upgrade CMake minimum version to 3.22 (#6916)

Fixes #6910.

alexreinking committed:
Remove unused GHA and packaging workflows. (#6917)

steven-johnson committed:
Fix two warnings found with clang 16 (#6918)

- variable 'count' set but not used
- warning: use of bitwise '|' with boolean operands

abadams committed:
Fix bug when realize condition depends on tuple call (#6915)

If the realization is tuple-valued, and the condition on the realization
uses a tuple call (index != 0), then the condition wasn't getting
resolved during the split_tuples pass. The cause was a missing mutate
call.

steven-johnson committed:
Fix wrong install path for *.py files (#6921)

* Fix wrong install path for *.py files

We were looking in a nonexistent dir, so we never copied `__init__.py` as we should have.

* Update CMakeLists.txt
